app/uk/gov/hmrc/gform/des/DesConnector.scala

```diff
@@ -31,17 +37,56 @@ import uk.gov.hmrc.http.logging.Authorization
 class DesConnector(wSHttp: WSHttp, baseUrl: String, desConfig: DesConnectorConfig) {
 
   def lookupRegistration(utr: String, desRegistrationRequest: DesRegistrationRequest)(
-    implicit ex: ExecutionContext): Future[DesRegistrationResponse] = {
+    implicit ex: ExecutionContext): Future[ServiceCallResponse[DesRegistrationResponse]] = {
+
     implicit val hc = HeaderCarrier(
       extraHeaders = Seq("Environment" -> desConfig.environment),
       authorization = Some(Authorization(s"Bearer ${desConfig.authorizationToken}")))
     Logger.info(s"Des registration, UTR: '$utr', ${loggingHelpers.cleanHeaderCarrierHeader(hc)}")
-    wSHttp.POST[DesRegistrationRequest, DesRegistrationResponse](
-      s"$baseUrl${desConfig.basePath}/registration/organisation/utr/$utr",
-      desRegistrationRequest)
 
+    wSHttp
+      .doPost[DesRegistrationRequest](
+        s"$baseUrl${desConfig.basePath}/registration/organisation/utr/$utr",
+        desRegistrationRequest,
+        List.empty[(String, String)]
+      )
+      .map { httpResponse =>
+        if (httpResponse.status === 400) {
+          processResponse[DesRegistrationResponseError, Nothing](httpResponse) { desError =>
+            if (desError.code === "INVALID_UTR") NotFound
+            else {
+              Logger.error("Problem when calling des registration: " + Json.prettyPrint(Json.toJson(desError)))
+              ServiceNotAvailable
+            }
+          }
+        } else processResponse[DesRegistrationResponse, DesRegistrationResponse](httpResponse)(ServiceResponse.apply)
+      }
+      .recover {
+        case ex =>
+          Logger.error("Unknown problem when calling des registration", ex)
+          ServiceNotAvailable
+      }
   }
 
+  private def processResponse[A: Format: TypeTag, B](httpResponse: HttpResponse)(
+    f: A => ServiceCallResponse[B]
+  ): ServiceCallResponse[B] =
+    Try(httpResponse.json) match {
+      case Success(jsValue) =>
+        jsValue.validate[A].map(f).recoverTotal { jsError =>
+          val tpe = implicitly[TypeTag[A]].tpe
+          Logger.error(s"Unknown problem when calling des registration. Expected json for $tpe, but got: $jsError")
+          ServiceNotAvailable
+        }
+      case Failure(failure) =>
+        val tpe = implicitly[TypeTag[A]].tpe
+        val status = httpResponse.status
+        Logger.error(
+          s"Unknown problem when calling des registration. Response status was: $status. Expected json for $tpe, but got :",
+          failure)
+        ServiceNotAvailable
```

Comment:
bad json does not mean ServiceNotAvailable, seems that this could give some confusion

Reply:
I agree with what you are saying, on the other hand bad json doesn't mean utr + postcode combination is invalid. As far as user is concerned we need to handle two cases.

1. utr + postcode combination is valid - ie we get response which we can turn into `DesRegistrationResponse`
2. utr + poscode combination is invalid - ie status=400 and we get response which we can turn into `DesRegistrationResponseError` and on top of that `response.code === "INVALID_UTR"`
3. any other case (being it wrong json, service being down etc.) we want to treat as technical difficulties (from user perspective).

To represent case 3. I'm using `ServiceNotAvailable` object, do you think renaming this object to `InvalidResponse` would be more appropriate? Any other ideas for name of that object?

Reply:
yes this is a bit confusing because of that, you can handle this on the frontend and return ServiceNotAvailable on that service but the backend should reflect what is actually happening as the user is not seeing that, and its easier for developers to diagnose